Hcsvlab 1019

shaunmuscat opened the pull request.

shaunmuscat committed:
HCSVLAB-1019: add api to add item metadata

Add api for adding items to a collection, which currently handles the addition of the metadata for a set of items. Does not provide ability to upload the item documents.

shaunmuscat committed:
HCSVLAB-1019: add document upload component to add item api

shaunmuscat committed:
HCSVLAB-1019: add handling of whitespace in name of uploaded files

The existing ingest code didn't appear to have any mechanisms in place to handle documents which have spaces in their file names. This is a small patch to enable the api to handle whitespace in filenames by subbing them with the '%20' character in the file reference URI. When the ingest looks for documents it now decodes the file path URI to properly decode the escaped whitespace before the file path is saved to the document record.

shaunmuscat committed:
HCSVLAB-1019: validate all item existence before doing document upload

lotophage commented on app/controllers/collections_controller.rb:
```diff
@@ -121,6 +121,93 @@ def revoke_access
     redirect_to licences_path
   end
 
+  def add_items_to_collection
```
This is *really* long method, which is what some people call a [code smell](https://en.wikipedia.org/wiki/Code_smell). Even though it is perfectly functional, it is probably doing a lot of work and a lot of different things, which makes it harder for other people to understand.

In this case, there is lots of validation code interspersed with processing code, you could probably move all the code to a `validate_add_items_request` type method. That method would itself just call different helper methods `validate_item`, `validate_document`, etc. which throw Exceptions with the appropriate message and response code.

The main block would then look something like:

```
def add_items_to_collection
  begin
    validate_add_items_request(params)
    ...
  rescue SomeException
    respond_with_error(e.message, e.response_code)
  end
end
```

lotophage commented on app/controllers/collections_controller.rb:
```diff
@@ -121,6 +121,93 @@ def revoke_access
+    elsif collection.nil?
+      respond_with_error("Requested collection not found", 404)
+    elsif params[:api_key] != User.find(collection.owner_id).authentication_token
+      respond_with_error("User is unauthorised", 403) # Authorise by comparing api key sent with collection owner's api key
```
I think the authorisation is already handled by the Devise gem so you don't need to worry about checking that.

shaun-intersect replied:
As far as I could determine, the gem authenticates that the api key provided belongs to an actual user, and that user is authorised to perform the actions they have requested (via the permissions file). 
This validation was intended to ensure that the owner of a collection is the only user that is authorised to call the api to add items to their collection.

lotophage replied:
Ahh, that makes sense!

lotophage commented on app/controllers/collections_controller.rb:
```diff
@@ -121,6 +121,93 @@ def revoke_access
+    # referenced documents (HCSVLAB-1019) are already handled by the look_for_documents part of the item ingest
+    collection = Collection.find_by_name(params[:id])
+    if !request.post? or params[:items].blank?
+      respond_with_error("JSON-LD formatted item metadata must be sent to the api call as a POST request", 400)
```
You can limit the types of requests that will be accepted in the `config/routes.rb`, so you won't need to check if it's a POST or not

lotophage commented on features/api.feature:
```diff
@@ -1487,4 +1487,253 @@ Feature: Browsing via API
\ No newline at end of file
+    """
+
+  #  @api_add_item
```
You probably don't need this test if you limit the routes. It's best to avoid putting commented code in the repo wherever possible because 99% of the time it never gets uncommented. If you create something that you *think* you might want to keep in the future, either create a new branch and commit it there, or stash it.

lotophage commented:
Lots of test cases, which is really good!

shaunmuscat committed:
HCSVLAB-1019: code cleanup

lotophage commented on app/controllers/collections_controller.rb:
```diff
@@ -189,4 +264,99 @@ def respond_with_error(message, status_code)
+  def validate_add_items_request(params)
+    collection = Collection.find_by_name(params[:id])
+    if params[:items].blank?
+      raise ResponseError.new(400), "JSON-LD formatted item metadata must be sent to the api call as a POST request"
```
Don't really need the POST request warning stuff since it is limited to a POST request anyway. It doesn't really make sense to send data via a GET request anyway.

lotophage commented on app/controllers/collections_controller.rb:
```diff
@@ -189,4 +264,99 @@ def respond_with_error(message, status_code)
+  # Validates the request on the add items api call
+  def validate_add_items_request(params)
+    collection = Collection.find_by_name(params[:id])
+    if params[:items].blank?
```
You could change the method signatures to all the validate methods to `validate_x(params)` for consistency. Then is method would mostly contain:
```
validate_collection(params)
validate_items(params)
validate_files(params)
```

All the iteration over objects can happen there

lotophage commented on app/controllers/collections_controller.rb:
```diff
@@ -189,4 +264,99 @@ def respond_with_error(message, status_code)
+    collection = Collection.find_by_name(params[:id])
+    if params[:items].blank?
+      raise ResponseError.new(400), "JSON-LD formatted item metadata must be sent to the api call as a POST request"
+    elsif collection.nil?
```
These collection existence and and whether the requester owns the collection could go into a `validate_collection(params)` method

lotophage commented on app/controllers/collections_controller.rb:
```diff
@@ -189,4 +264,99 @@ def respond_with_error(message, status_code)
+    elsif params[:api_key] != User.find(collection.owner_id).authentication_token
+      raise ResponseError.new(403), "User is unauthorised" # Authorise by comparing api key sent with collection owner's api key
+    end
+    params[:items].each do |item|
```
this iteration can be pushed down into the `validate_items` method, and the document iteration can be pushed into `validate_documents`

lotophage commented on app/controllers/collections_controller.rb:
```diff
@@ -121,6 +121,60 @@ def revoke_access
+    # referenced documents (HCSVLAB-1019) are already handled by the look_for_documents part of the item ingest
+    begin
+      # Perform parameter transformations and request validations
+      params[:items] = JSON.parse(params[:items]) if params[:items].is_a? String
```
These two steps could go into some sort of cleanse params type method

lotophage commented on app/controllers/collections_controller.rb:
```diff
@@ -189,4 +264,99 @@ def respond_with_error(message, status_code)
+  # Uploads a document given as a http multipart uploaded file or responds with an error if appropriate
+  def upload_document_using_multipart(absolute_filename, file, collection_name)
+    filename = File.basename absolute_filename
+    if !file.is_a? ActionDispatch::Http::UploadedFile
```
try to move all this validation code into file validation step

lotophage commented on app/controllers/collections_controller.rb:
```diff
@@ -121,6 +121,60 @@ def revoke_access
+      # Upload each of the files provided in the multipart HTTP request
+      if !params[:file].nil?
+        params[:file].each do |uploaded_file|
+          uploaded_files.push(upload_document_using_multipart(File.join(corpus_dir, uploaded_file.original_filename), uploaded_file, collection.name))
```
You call `File.join(corpus_dir, uploaded_file.original_filename)` here, but he first line `upload_document_using_multipart` is `filename = File.basename absolute_filename`, it seems a bit funny to create a file path then unpack it again straight away. Maybe pass the corpus dir and the filename as separate params to `upload_document_using_multipart` and do the `File.join` there, that way you don't ever have to call `File.basename`

lotophage commented on app/controllers/collections_controller.rb:
```diff
@@ -189,4 +264,99 @@ def respond_with_error(message, status_code)
 
+  # Uploads a document given as json content or responds with an error if appropriate
+  def upload_document_using_json(filename, json_content, collection_name)
+    if File.exists? filename
```
This could also be moved to the file validation method

lotophage commented on app/controllers/collections_controller.rb:
```diff
@@ -121,6 +121,60 @@ def revoke_access
+        end
+      end
+
+      # Update document source in graph, convert metadata to RDF and ingest each item
```
This method is still doing a lot of work, you could split the rest into a few extra methods, 
```
def add_items_to_collection
  begin
    cleans_params(params)
    validate_add_items_request(params)
    process_uploaded_files
    process_items # which probably has a process_documents helper
    ingest_items
resuce
    ...
```
If you have good names for variables and methods, the advantage of having lots of little methods over one big one is that it reads like pseudocode, so it is very easy for someone else to understand. Even if it was completely broken, someone else can come along and understand what the intent was. It also makes it easier to track down bugs because they can be isolated to smaller sections of code.

lotophage commented:
I know I've made lots of comments, but overall the code is looking great, and is already a million times easier to understand. Always remember that my remarks are my *opinions* so you are always welcome to disagree!

shaunmuscat committed:
HCSVLAB-1019: code refactor

shaun-intersect commented:
Suggestions given by Dan to:
- validate and catch case when malformed JSON is given in the request, it appears that this is currently unhandled.
- perform all validations upfront before doing any program logic, including the validations that uploaded files don't already exist on the server.
- provide some warning message if files already exist on the server (if it is meant to skip existing files rather than exit/fail when one is encountered).

shaunmuscat committed:
HCSVLAB-1019: modify code to handle new test cases

New test cases added to handle invalid JSON for item metadata and identical file identifiers

shaunmuscat committed:
HCSVLAB-1019: code refactor

lotophage commented on app/controllers/collections_controller.rb:
```diff
@@ -189,4 +233,218 @@ def respond_with_error(message, status_code)
+
+  # Returns a cleansed copy of params for the add item api
+  def cleanse_params(request_params)
+    unless request_params.nil?
```
You are a very defensive programmer which is good, but you are perhaps overly cautious and you are doing lots of unnecessary checks. Can `params` ever be `nil` or will it be an empty hash?

Here you don't need to check if it's a string because `JSON.parse` will do that for you. You don't really need the `valid_json?` helper method because `JSON.parse` will throw an error if it isn't. You just need to have a begin/rescue block here that catches the JSON parse error and re-raises it as a response `ResponseError` that way you don't need the helper and only call `JSON.parse` once instead of twice.

shaun-intersect replied:
The string check is required because the `add_items_to_collection` method could be provided with either a String or an actual Array or Hash, depending on how the api request was posted. This is significant because calling `JSON.parse` on something which has already been parsed results in an error similar to the following `TypeError (no implicit conversion of Array into String)`.

This can be seen when running the feature tests with the proposed change. A more real-case example is that if the request contains the header `"Content-Type: application/json"` then somewhere in the Rails app `param[:items]` is automatically converted into an array/hash, however if this header is left out (such as when making a multipost request) then `param[:items]` remains a string and needs to be parsed into JSON.

lotophage replied:
Fair enough

lotophage commented on app/controllers/collections_controller.rb:
```diff
@@ -189,4 +233,218 @@ def respond_with_error(message, status_code)
+        raise ResponseError.new(400), "JSON item metadata is ill-formatted" unless valid_json?(request_params[:items])
+        request_params[:items] = JSON.parse(request_params[:items])
+      end
+      request_params[:file] = [request_params[:file]] unless request_params[:file].nil? or request_params[:file].is_a? Array
```
Returning empty collections in general is better than leaving something `nil` because it simplifies iteration code further down that track. Later on you can do:
```
files.each do |file|
...
end
```

instead of:
```
unless files.nil?
  files.each do |file|
  ...
  end
end

```

lotophage commented on app/controllers/collections_controller.rb:
```diff
@@ -189,4 +233,218 @@ def respond_with_error(message, status_code)
+  # Updates the metadata graph for the given item
+  # Returns a copy of the item with the document sourcs in the graph updates to the path of an uploaded file when appropriate
+  def update_item_graph_with_uploaded_files(uploaded_files, item_metadata)
+    unless uploaded_files.nil?
```
You don't need this wrapped in a `nil` check because your `process_uploaded_files` will never return a `nil`, it will return an empty array (which is good!)

lotophage commented on app/controllers/collections_controller.rb:
```diff
@@ -189,4 +233,218 @@ def respond_with_error(message, status_code)
+        end
+      end
+    end
+    unless uploaded_files.nil?
```
I don't think you need this check because `process_uploaded_files` will always return an array (possibly empty)

lotophage commented on app/controllers/collections_controller.rb:
```diff
@@ -189,4 +233,218 @@ def respond_with_error(message, status_code)
+
+  # Validates each of the uploaded files
+  def validate_files(uploaded_files, collection, corpus_dir)
+    unless uploaded_files.nil?
```
could possible avoid this nil check if `cleanse_params` replaces a `params[:file]` param with an empty array

lotophage commented on app/controllers/collections_controller.rb:
```diff
@@ -121,6 +121,29 @@ def revoke_access
+      corpus_dir = corpus_dir(request_params[:id])
+      validate_add_items_request(collection, corpus_dir, request_params)
+      uploaded_files = process_uploaded_files(corpus_dir, collection.name, request_params[:file])
+      items = []
```
everything from `items = []` up to and including `raise ResponseError.new(400), "No items were added" if items.blank?` would be better off in `process_items` helper method

lotophage commented:
Looking really good. A few minor nitpicks, but otherwise your code is starting to look really clean :sparkles:

shaunmuscat committed:
HCSVLAB-1019: refactor code

Refactor response exception and request validation out into lib module

lotophage commented:
:+1: merge
